infrastructure/src/main/java/org/corfudb/infrastructure/LogUnitServer.java

```diff
@@ -131,69 +139,120 @@ public LogUnitServer(ServerContext serverContext) {
     }
 
+    //TODO(Maithem): set LU thread pool, then accept vs then accept async?
+
     /**
      * Service an incoming request to retrieve the starting address of this logging unit.
      */
     @ServerHandler(type = CorfuMsgType.TRIM_MARK_REQUEST)
     public void handleHeadRequest(CorfuMsg msg, ChannelHandlerContext ctx, IServerRouter r) {
+        // TODO(Maithem): Should go through the batchwriter?
         r.sendResponse(ctx, msg, CorfuMsgType.TRIM_MARK_RESPONSE.payloadMsg(streamLog.getTrimMark()));
     }
 
+    // TODO(Maithem): converting exceptions to message types is really superfluous,
+    // we should be able to send the exception directly
+    void handleWriteException(Throwable ex, ChannelHandlerContext ctx, CorfuPayloadMsg msg, IServerRouter r) {
+        if (ex.getCause() instanceof OverwriteException) {
+            OverwriteException owe = (OverwriteException) ex.getCause();
+            r.sendResponse(ctx, msg, CorfuMsgType.ERROR_OVERWRITE
+                    .payloadMsg(owe.getOverWriteCause().getId()));
+        } else if (ex.getCause() instanceof DataOutrankedException) {
+            r.sendResponse(ctx, msg, CorfuMsgType.ERROR_DATA_OUTRANKED.msg());
+        } else if (ex.getCause() instanceof ValueAdoptedException) {
+            ValueAdoptedException vae = (ValueAdoptedException) ex.getCause();
+            r.sendResponse(ctx, msg, CorfuMsgType.ERROR_VALUE_ADOPTED.payloadMsg(vae.getReadResponse()));
+        } else if (ex.getCause() instanceof TrimmedException) {
+            r.sendResponse(ctx, msg, CorfuMsgType.ERROR_TRIMMED.msg());
+        } else {
+            throw new LogUnitException(ex);
+        }
+    }
+
     /**
      * Service an incoming write request.
      */
     @ServerHandler(type = CorfuMsgType.WRITE)
     public void write(CorfuPayloadMsg<WriteRequest> msg, ChannelHandlerContext ctx, IServerRouter r) {
         log.debug("log write: global: {}, streams: {}", msg.getPayload().getToken(),
                 msg.getPayload().getData().getBackpointerMap());
-
-        try {
-            LogData logData = (LogData) msg.getPayload().getData();
-            logData.setEpoch(msg.getEpoch());
-            dataCache.put(msg.getPayload().getGlobalAddress(), logData);
-            r.sendResponse(ctx, msg, CorfuMsgType.WRITE_OK.msg());
-
-        } catch (OverwriteException ex) {
-            r.sendResponse(ctx, msg, CorfuMsgType.ERROR_OVERWRITE.payloadMsg(ex.getOverWriteCause().getId()));
-        } catch (DataOutrankedException e) {
-            r.sendResponse(ctx, msg, CorfuMsgType.ERROR_DATA_OUTRANKED.msg());
-        } catch (ValueAdoptedException e) {
-            r.sendResponse(ctx, msg, CorfuMsgType.ERROR_VALUE_ADOPTED.payloadMsg(e
-                    .getReadResponse()));
-        }
+        LogData logData = (LogData) msg.getPayload().getData();
+
+        batchWriter.addTask(WRITE, msg)
+                .thenAccept(Void -> {
```

Comment:
thenAccept might block the main thread.

Reply:
you can use `thenRun(..)` instead of `thenAccept()` it fits better

Reply:
Done.